pyfifotax/utils.py

```diff
@@ -0,0 +1,548 @@
+import os
+import io
+import datetime
+import requests
+import zipfile
+import decimal
+import warnings
+from dataclasses import dataclass
+from datetime import timedelta
+
+import pandas as pd
+
+
+def get_date(forex):
+    return forex.date
+
+
+def round_decimal(number, precision: str = "0.01"):
+    return number.quantize(
+        decimal.Decimal(precision),
+        rounding=decimal.ROUND_HALF_UP,
+    )
+
+
+def to_decimal(number):
+    return decimal.Decimal(number)
+
+
+def sum_decimal(series: pd.Series):
+    if series.shape[0] == 0:
+        return to_decimal(0)
+    return series.sum()
+
+
+def get_reference_rates():
+    # check whether to download more recent exchange-rate data
+    mod_date = None
+    today = datetime.date.today()
+    if os.path.exists("eurofxref-hist.csv"):
+        mod_time = os.path.getmtime("eurofxref-hist.csv")
+        mod_date = datetime.datetime.fromtimestamp(mod_time).date()
+
+    if mod_date != today:
+        print("Downloading more recent exchange rate data ...")
+        response = requests.get(
+            "https://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist.zip?69474b1034fa15ae36103fbf3554d272"
+        )
+        zip_file = zipfile.ZipFile(io.BytesIO(response.content))
+        zip_file.extractall(".")
+
+    daily_ex_rates = pd.read_csv("eurofxref-hist.csv", parse_dates=["Date"])
+    daily_ex_rates = daily_ex_rates.loc[
+        :, ~daily_ex_rates.columns.str.contains("^Unnamed")
+    ]
+
+    daily_ex_rates.index = daily_ex_rates["Date"]
+    # drop years earlier as 2009 as this would make reporting tax earning
+    # way more complicated anyway
+    daily_ex_rates = daily_ex_rates.loc[daily_ex_rates.index.year >= 2009]
+    # drop columns with nan values
+    daily_ex_rates = daily_ex_rates.dropna(axis="columns")
+
+    daily_ex_rates = daily_ex_rates.drop("Date", axis="columns")
+    daily_ex_rates = daily_ex_rates
+    monthly_ex_rates = daily_ex_rates.groupby(
+        by=[daily_ex_rates.index.year, daily_ex_rates.index.month]
+    ).mean()
+
+    supported_currencies = set(daily_ex_rates.columns)
+    supported_currencies.add("EUR")
+
+    return daily_ex_rates, monthly_ex_rates, supported_currencies
+
+
+def get_monthly_rate(monthly_rates: pd.DataFrame, date: datetime, currency: str):
+    return to_decimal(monthly_rates[currency][date.year, date.month])
+
+
+def get_daily_rate(daily_rates: pd.DataFrame, date: datetime, currency: str):
+    if date in daily_rates[currency]:
```

Comment:
I wanted to add EUR to my balance to the Forex sheet, and did it the way that I "converted" the desired amount from EUR to EUR.

However, in this case this errs out here with `KeyError('EUR')`.

Reply:
Good catch, I'll modify the application of exchange rates such that the "exchange rate" is 1 if the currency equals the domestic currency or equals EUR for AWV purposes.